central/hash/manager/manager.go

```diff
@@ -0,0 +1,128 @@
+package manager
+
+import (
+	"context"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/stackrox/rox/central/hash/datastore"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/concurrency"
+	"github.com/stackrox/rox/pkg/logging"
+	"github.com/stackrox/rox/pkg/sync"
+)
+
+const (
+	flushInterval = 1 * time.Minute
+)
+
+var (
+	log = logging.LoggerForModule()
+)
+
+// Manager is a hash manager that provides access to cluster-based dedupers and persists
+// the hashes into the database
+type Manager interface {
+	Start(ctx context.Context)
+
+	GetDeduper(ctx context.Context, clusterID string) Deduper
+	Delete(ctx context.Context, clusterID string) error
+}
+
+// NewManager instantiates a Manager
+func NewManager(datastore datastore.Datastore) Manager {
+	return &managerImpl{
+		datastore: datastore,
+		dedupers:  make(map[string]Deduper),
+	}
+}
+
+type managerImpl struct {
+	datastore datastore.Datastore
+
+	dedupersLock sync.RWMutex
+	dedupers     map[string]Deduper
+}
+
+func (m *managerImpl) flushHashes(ctx context.Context) {
+	// Get clusters first to flush hashes one at a time
+	var clusters []string
+	concurrency.WithLock(&m.dedupersLock, func() {
+		clusters = make([]string, 0, len(m.dedupers))
+		for clusterID := range m.dedupers {
+			clusters = append(clusters, clusterID)
+		}
+	})
+	for _, cluster := range clusters {
+		var deduper Deduper
+		var ok bool
+		concurrency.WithLock(&m.dedupersLock, func() {
+			deduper, ok = m.dedupers[cluster]
+		})
+		if !ok {
+			continue
+		}
+		hash := &storage.Hash{
+			ClusterId: cluster,
+			Hashes:    deduper.GetSuccessfulHashes(),
+		}
+		if err := m.datastore.UpsertHash(ctx, hash); err != nil {
+			log.Errorf("flushing hashes: %v", err)
+		}
+		dedupingHashSizeGauge.With(prometheus.Labels{"cluster": hash.GetClusterId()}).Set(float64(len(hash.GetHashes())))
+	}
+}
+
+func (m *managerImpl) Start(ctx context.Context) {
+	t := time.NewTicker(flushInterval)
+	defer t.Stop()
+	for {
+		select {
+		case <-t.C:
+			m.flushHashes(ctx)
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (m *managerImpl) getDeduper(clusterID string) (Deduper, bool) {
+	m.dedupersLock.RLock()
+	defer m.dedupersLock.RUnlock()
+
+	d, ok := m.dedupers[clusterID]
+	return d, ok
+}
+
+func (m *managerImpl) GetDeduper(ctx context.Context, clusterID string) Deduper {
+	d, ok := m.getDeduper(clusterID)
+	if ok {
+		return d
+	}
+	hash, exists, err := m.datastore.GetHashes(ctx, clusterID)
+	if err != nil {
+		log.Errorf("could not get hashes from database for cluster %q: %v", clusterID, err)
+	}
+	if !exists {
+		d = NewDeduper(make(map[string]uint64))
+	} else {
+		d = NewDeduper(hash.GetHashes())
+	}
+
+	m.dedupersLock.Lock()
+	defer m.dedupersLock.Unlock()
+
+	existingDeduper, ok := m.dedupers[clusterID]
+	if ok {
```

Comment:
I assume this case only ever happens if two secured clusters with the same bundle attempt to connect to Central at the same time? 

To avoid having to check here again if `m.deduper` has an entry for this cluster, why not make the entire function a critical section? This way you'd be sure that if it wasn't set in the beginning of this function, you can safely set it to the new deduper created. I don't think this is called often enough to become a contention problem.

Reply:
Yeah the concern was wrapping around a call to postgres, and blocking other calls to not occur in parallel